### ISSUE TITLE
Upgrade to hollodotme/fast-cgi-client v3.0

### ISSUE THREAD

**Issue by mnapoli** [brefphp/bref#388]
A 3.0 has been released for the FastCGI client we are using: https://github.com/hollodotme/fast-cgi-client/releases/tag/v3.0.0

We should upgrade to this version. It might also help to handle multi-value headers, which was discussed here: https://github.com/brefphp/bref/pull/319#issuecomment-487773479 and #213

**Comment by linuxjuggler** [comment]
Hello @mnapoli

I am working on the upgrade, but I've noticed that based on https://github.com/hollodotme/fast-cgi-client/blob/master/CHANGELOG.md#300-alpha---2019-04-30 the headers will always return an array which caused the following tests to fail

1. https://github.com/brefphp/bref/blob/master/tests/Runtime/PhpFpmTest.php#L782
2. https://github.com/brefphp/bref/blob/master/tests/Runtime/PhpFpmTest.php#L811

Fixing them will make them a duplicate of:

1. test response with multivalue headers
2. test response with multiple cookies with multiheader

So do you advice to remove them? Or you have an idea how should I update them?

Thanks

**Comment by mnapoli** [comment]
@linuxjuggler sorry I'm having a doubt right now: are multi-value headers always enabled? What I mean is that is the `multiValueHeaders` field always present in the AWS Lambda event?

Because if it is, we could always use it, and always return multi-value headers. In that case it could make sense to remove those tests.

If not, we need to support 2 modes: multi-value headers mode, and non-multi-value headers mode. In that case we need to keep these tests as they test the case where the `multiValueHeaders` field doesn't exist.

**Comment by linuxjuggler** [comment]
from what I can tell, and maybe am wrong, what they mention [here](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#multi-value-headers) is that multi-value headers are not enabled by default, instead, we will have to [enable](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#enable-multi-value-headers) them manually.

I guess then we will have to support the two modes by, and correct me if I am wrong, modifying the get response to either return a string or an array instead of returning an array.

**Comment by mnapoli** [comment]
The links you gave are specific to ALB (not API Gateway). But in any case that makes sense, and even that piece of code adds to the fact that multi-value headers are not always there:

https://github.com/brefphp/bref/blob/b7db4c4bdf4da8727c5301b6f782154706e836ea/runtime/layers/fpm/bootstrap#L32-L33

So you need to keep the tests that cover the case where there is no multi-value header.

Keep in mind that there are 2 separate things:

- the FastCGI client that can deal with multi-value headers
- the event that deals with multi-value headers

With v3.0 IIRC we will always have multi-value headers when talking with PHP-FPM. But that doesn't mean that this is the same for the events. Which is why we need to keep high-level tests (the one where we send an event) unaware of what's happening in the internals of Bref (the fastcgi communication). Hope that makes sense!

**Comment by linuxjuggler** [comment]
Yes it does.

I am looking at this [Set up Lambda Proxy Integrations in API Gateway ](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#apigateway-multivalue-headers-and-parameters), and this is what they mentioned:

```
API Gateway supports multiple headers and query string parameters that have the same name. Multi-value headers, as well as single-value headers and parameters, can be combined in the same requests and responses.
```

So by default, the multi-value header is enabled in the API Gateway, but it needs to be enabled manually in the ALB.

I'll modify my code and see how to get the tests to pass.

Thanks @mnapoli and sorry for any inconvenience

**Comment by mnapoli** [comment]
no worries thank you for doing this! And yep let's keep support for both cases then, I know some people use ALB.

**Comment by linuxjuggler** [comment]
@mnapoli can you check https://github.com/linuxjuggler/bref/commit/10b0bad568c17fc60c7b64aef4996cfa15f19f40 and let me know if you have any comments

all tests passed without any modifications https://travis-ci.org/linuxjuggler/bref/builds/569795721

**Comment by mnapoli** [comment]
Looking good! A few questions:

- is a connection to PHP-FPM recreated for each HTTP request? That might be costly
- what is the purpose of `array_values(array_slice($value, -1))[0]`?

Feel free to open a pull request, it may be easier to discuss this. Also we may be able to ping the library author to get his advice.

**Comment by linuxjuggler** [comment]
> is a connection to PHP-FPM recreated for each HTTP request? That might be costly

I don't think so, am not so familiar with how `fast-cgi-client` works tbh, I just did the modifications but didn't look deep on that part.

> what is the purpose of array_values(array_slice($value, -1))[0]?

To return the last header value as text, since as mentioned in https://github.com/hollodotme/fast-cgi-client/blob/master/CHANGELOG.md#300-alpha---2019-04-30 there is a BC break as the headers now will always return as an array instead of a string 

> Method Response#getHeaders() : array will now return a two-dimensional array with grouped values to support multi-value headers. Previous versions returned a one-dimensional key-value array.

```
# Previous versions
[
  'Status' => 'HTTP/2 200 OK',
  'Set-Cookie' => 'tasty_cookie=strawberry',
]

# Since 3.0.0-alpha
[
  'Status' => [
    'HTTP/2 200 OK',
  ],
  'Set-Cookie' => [
    'yummy_cookie=choco',
    'tasty_cookie=strawberry',
  ],
]
```

> Method Response#getHeader(string $headerKey) : array will now return an array containing all values for the given header key to support multi-value headers. Previous versions returned the last value as a string.

```
echo $response->getHeader('Set-Cookie');

# Previous versions
'tasty_cookie=strawberry'

# Since 3.0.0-alpha
[
  'yummy_cookie=choco',
  'tasty_cookie=strawberry'
]
```

I'll create a pull request in few min

**Comment by mnapoli** [comment]
Fixed in #405
